Re-diffed for 2.6.14.7 by Kerin Millar <kerframil@gmail.com>:

From: Chris Wright <chrisw@sous-sol.org>
Date: Fri, 17 Feb 2006 21:59:36 +0000 (-0800)
Subject: [PATCH] sys_mbind sanity checking
X-Git-Url: http://www.kernel.org/git/?p=linux/kernel/git/chrisw/linux-2.6.15.y.git;a=commitdiff;h=99e1baf86ff6a31330ff0c7bb77110338f03ddc5

[PATCH] sys_mbind sanity checking

Make sure maxnodes is safe size before calculating nlongs in
get_nodes().

Signed-off-by: Chris Wright <chrisw@sous-sol.org>
Signed-off-by: Linus Torvalds <torvalds@osdl.org>
[chrisw: fix units, pointed out by Andi]
Cc: Andi Kleen <ak@suse.de>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---

--- linux-2.6.14.7.orig/mm/mempolicy.c	2005-10-27 19:02:08.000000000 -0500
+++ linux-2.6.14.7/mm/mempolicy.c	2006-03-08 03:01:55.000000000 -0500
@@ -139,6 +139,8 @@ static int get_nodes(unsigned long *node
 	bitmap_zero(nodes, MAX_NUMNODES);
 	if (maxnode == 0 || !nmask)
 		return 0;
+	if (maxnode > PAGE_SIZE*BITS_PER_BYTE)
+		return -EINVAL;
 
 	nlongs = BITS_TO_LONGS(maxnode);
 	if ((maxnode % BITS_PER_LONG) == 0)
